diff --git a/Sources/CNELwIP/src/core/ipv4/ip4.c b/Sources/CNELwIP/src/core/ipv4/ip4.c
index e044bff..c653007 100644
--- a/Sources/CNELwIP/src/core/ipv4/ip4.c
+++ b/Sources/CNELwIP/src/core/ipv4/ip4.c
@@ -213,16 +213,17 @@ ip4_route(const ip4_addr_t *dest)
 #endif
 #endif /* !LWIP_SINGLE_NETIF */

-  if ((netif_default == NULL) || !netif_is_up(netif_default) || !netif_is_link_up(netif_default) ||
-      ip4_addr_isany_val(*netif_ip4_addr(netif_default)) || ip4_addr_isloopback(dest)) {
-    /* No matching netif found and default netif is not usable.
-       If this is not good enough for you, use LWIP_HOOK_IP4_ROUTE() */
-    LWIP_DEBUGF(IP_DEBUG | LWIP_DBG_LEVEL_SERIOUS, ("ip4_route: No route to %"U16_F".%"U16_F".%"U16_F".%"U16_F"\n",
-                ip4_addr1_16(dest), ip4_addr2_16(dest), ip4_addr3_16(dest), ip4_addr4_16(dest)));
-    IP_STATS_INC(ip.rterr);
-    MIB2_STATS_INC(mib2.ipoutnoroutes);
-    return NULL;
-  }
+  /* CNELwIP - Always use netif_default to route IPv4 traffic. */
+  //  if ((netif_default == NULL) || !netif_is_up(netif_default) || !netif_is_link_up(netif_default) ||
+  //      ip4_addr_isany_val(*netif_ip4_addr(netif_default)) || ip4_addr_isloopback(dest)) {
+  //    /* No matching netif found and default netif is not usable.
+  //       If this is not good enough for you, use LWIP_HOOK_IP4_ROUTE() */
+  //    LWIP_DEBUGF(IP_DEBUG | LWIP_DBG_LEVEL_SERIOUS, ("ip4_route: No route to %"U16_F".%"U16_F".%"U16_F".%"U16_F"\n",
+  //                ip4_addr1_16(dest), ip4_addr2_16(dest), ip4_addr3_16(dest), ip4_addr4_16(dest)));
+  //    IP_STATS_INC(ip.rterr);
+  //    MIB2_STATS_INC(mib2.ipoutnoroutes);
+  //    return NULL;
+  //  }

   return netif_default;
 }
@@ -438,6 +439,11 @@ ip4_input_accept(struct netif *netif)
       return 1;
     }
 #endif /* LWIP_AUTOIP */
+
+    /* CNELwIP - Our netif Accept ALL packets. */
+    LWIP_DEBUGF(IP_DEBUG, ("ip4_input: CNELwIP ACCEPT ALL packets on interface %c%c\n",
+                           netif->name[0], netif->name[1]));
+    return 1;
   }
   return 0;
 }
diff --git a/Sources/CNELwIP/src/core/tcp_in.c b/Sources/CNELwIP/src/core/tcp_in.c
index 1b17e40..9d7959c 100644
--- a/Sources/CNELwIP/src/core/tcp_in.c
+++ b/Sources/CNELwIP/src/core/tcp_in.c
@@ -315,40 +315,43 @@ tcp_input(struct pbuf *p, struct netif *inp)
     /* Finally, if we still did not get a match, we check all PCBs that
        are LISTENing for incoming connections. */
     prev = NULL;
-    for (lpcb = tcp_listen_pcbs.listen_pcbs; lpcb != NULL; lpcb = lpcb->next) {
-      /* check if PCB is bound to specific netif */
-      if ((lpcb->netif_idx != NETIF_NO_INDEX) &&
-          (lpcb->netif_idx != netif_get_index(ip_data.current_input_netif))) {
-        prev = (struct tcp_pcb *)lpcb;
-        continue;
-      }

-      if (lpcb->local_port == tcphdr->dest) {
-        if (IP_IS_ANY_TYPE_VAL(lpcb->local_ip)) {
-          /* found an ANY TYPE (IPv4/IPv6) match */
-#if SO_REUSE
-          lpcb_any = lpcb;
-          lpcb_prev = prev;
-#else /* SO_REUSE */
-          break;
-#endif /* SO_REUSE */
-        } else if (IP_ADDR_PCB_VERSION_MATCH_EXACT(lpcb, ip_current_dest_addr())) {
-          if (ip_addr_eq(&lpcb->local_ip, ip_current_dest_addr())) {
-            /* found an exact match */
-            break;
-          } else if (ip_addr_isany(&lpcb->local_ip)) {
-            /* found an ANY-match */
-#if SO_REUSE
-            lpcb_any = lpcb;
-            lpcb_prev = prev;
-#else /* SO_REUSE */
-            break;
-#endif /* SO_REUSE */
-          }
-        }
-      }
-      prev = (struct tcp_pcb *)lpcb;
-    }
+    /* CNELwIP - Our pcb listen on ALL address and ALL ports. */
+    lpcb = tcp_listen_pcbs.listen_pcbs;
+//    for (lpcb = tcp_listen_pcbs.listen_pcbs; lpcb != NULL; lpcb = lpcb->next) {
+//      /* check if PCB is bound to specific netif */
+//      if ((lpcb->netif_idx != NETIF_NO_INDEX) &&
+//          (lpcb->netif_idx != netif_get_index(ip_data.current_input_netif))) {
+//        prev = (struct tcp_pcb *)lpcb;
+//        continue;
+//      }
+//
+//      if (lpcb->local_port == tcphdr->dest) {
+//        if (IP_IS_ANY_TYPE_VAL(lpcb->local_ip)) {
+//          /* found an ANY TYPE (IPv4/IPv6) match */
+//#if SO_REUSE
+//          lpcb_any = lpcb;
+//          lpcb_prev = prev;
+//#else /* SO_REUSE */
+//          break;
+//#endif /* SO_REUSE */
+//        } else if (IP_ADDR_PCB_VERSION_MATCH_EXACT(lpcb, ip_current_dest_addr())) {
+//          if (ip_addr_eq(&lpcb->local_ip, ip_current_dest_addr())) {
+//            /* found an exact match */
+//            break;
+//          } else if (ip_addr_isany(&lpcb->local_ip)) {
+//            /* found an ANY-match */
+//#if SO_REUSE
+//            lpcb_any = lpcb;
+//            lpcb_prev = prev;
+//#else /* SO_REUSE */
+//            break;
+//#endif /* SO_REUSE */
+//          }
+//        }
+//      }
+//      prev = (struct tcp_pcb *)lpcb;
+//    }
 #if SO_REUSE
     /* first try specific local IP */
     if (lpcb == NULL) {
@@ -675,7 +678,11 @@ tcp_listen_input(struct tcp_pcb_listen *pcb)
     /* Set up the new PCB. */
     ip_addr_copy(npcb->local_ip, *ip_current_dest_addr());
     ip_addr_copy(npcb->remote_ip, *ip_current_src_addr());
-    npcb->local_port = pcb->local_port;
+    /* CNELwIP - Modify local port.
+     Imaging that we are running a virtual server handles all request
+     from localhost, so the local_ip is actually the remote server
+     ip (since this stack is running as a stack of remote server). */
+    npcb->local_port = tcphdr->dest;
     npcb->remote_port = tcphdr->src;
     npcb->state = SYN_RCVD;
     npcb->rcv_nxt = seqno + 1;
